### PR TITLE
Use luaossl instead of luacrypto

### PR DESCRIPTION
**Pull request by coderforlife**
LuaCrypto is very difficult to install on modern releases of Ubuntu (and likely other OSes) due to it being incompatible with newer libopenssl versions and no longer maintained. This PR simply replaces its usage with the actively maintained luaossl library. This does not change the usage of the library at all, only modifies the internal library used.

The [jitsi/luajwt](https://github.com/jitsi/luajwt) repo is still needed by https://github.com/jitsi/jitsi-meet/blob/master/resources/prosody-plugins/token/util.lib.lua which is part of the recommended way of using tokens with JWT.

There seems to be some code in [jitsi/lua-jwt](https://github.com/jitsi/lua-jwt) (notice the hyphen there) that does use luaossl instead of LuaCrypto, however that library is not a drop-in replacement in the prosody-plugins and I have no idea where it is used. Additionally, it has not been updated for 4 years as well.

### CODE REVIEW

**Comment by coderforlife** [comment]
Just noticed the other two pull requests. There should be no substantial differences between them except:

- They include tests
- I updated the lua-cjson dependency which causes problems

File: luajwtjitsi-1.3-7.rockspec [hunk 1]
```diff
@@ -15,8 +15,8 @@ description = {
 
 dependencies = {
 	"lua >= 5.1",
-	"luacrypto >= 0.3.2-1",
-	"lua-cjson >= 2.1.0",
+	"luaossl >= 20151221-0",
+	"lua-cjson-ol >= 1.0-1",
```
Comment (saghul) [inline luajwtjitsi-1.3-7.rockspec:19]: why is this change necessary?

**Comment by coderforlife** [comment]
Currently if you try doing `luarocks install lua-cjson >= 2.1.0` you get a different library (that causes issues) than expected. The lua-cjson-ol is compatible with Lua 5.2 / Prosody 11 which all semi-recent OSes use. See https://github.com/mpx/lua-cjson/issues/56#issuecomment-394764240 and https://github.com/mpx/lua-cjson/issues/56#issuecomment-401064789 for some other info.

Also, It is what is used by [jitsi/lua-jwt](https://github.com/jitsi/lua-jwt) (with a hyphen).

Another option to fix it would be to just use `= 2.1.0` instead of `>=2.1.0` to make sure the newer, incompatible, version is not installed.

**Comment by saghul** [comment]
Here is how we've been installing it on the Docker setup: https://github.com/jitsi/docker-jitsi-meet/blob/a7563d4f82d79cfaca6f5ff408299f7857f7cac7/prosody/Dockerfile#L16 I'd say we pin the dependency to that, since it's what is "tried and true".

**Comment by coderforlife** [comment]
Pinning the version would also work fine (still new with lua, so I don't know if that is `= 2.1.0` or just `2.1.0`).

**Comment by MobtgZhang** [comment]
Hello,I am using torch library,but when installing iTorch , the library is missed dependencies for itorch: luacrypto. Therefore,I decide to install luacrypto. But how I install luacrypto with different versions of gcc compiler(gcc-9,gcc-7,gcc-4.8), the compiler always reports an error for the following message:

`gcc -O2 -fPIC -I/home/mobtgzhang/torch/install/include -c src/lcrypto.c -o src/lcrypto.o -I/usr/include
src/lcrypto.c: In function ‘digest_pnew’:
src/lcrypto.c:81:61: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_MD_CTX’ {aka ‘struct evp_md_ctx_st’}
   81 |     EVP_MD_CTX *c = (EVP_MD_CTX *)lua_newuserdata(L, sizeof(EVP_MD_CTX));
      |                                                             ^~~~~~~~~~
src/lcrypto.c: In function ‘digest_reset’:
src/lcrypto.c:120:10: warning: implicit declaration of function ‘EVP_MD_CTX_cleanup’; did you mean ‘EVP_MD_CTX_create’? [-Wimplicit-function-declaration]
  120 |     if (!EVP_MD_CTX_cleanup(c))
      |          ^~~~~~~~~~~~~~~~~~
      |          EVP_MD_CTX_create
src/lcrypto.c: In function ‘encrypt_pnew’:
src/lcrypto.c:331:69: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_CIPHER_CTX’ {aka ‘struct evp_cipher_ctx_st’}
  331 |     EVP_CIPHER_CTX *c = (EVP_CIPHER_CTX *)lua_newuserdata(L, sizeof(EVP_CIPHER_CTX));
      |                                                                     ^~~~~~~~~~~~~~
src/lcrypto.c: In function ‘encrypt_fencrypt’:
src/lcrypto.c:425:20: error: storage size of ‘c’ isn’t known
  425 |     EVP_CIPHER_CTX c;
      |                    ^
src/lcrypto.c: In function ‘decrypt_pnew’:
src/lcrypto.c:470:69: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_CIPHER_CTX’ {aka ‘struct evp_cipher_ctx_st’}
  470 |     EVP_CIPHER_CTX *c = (EVP_CIPHER_CTX *)lua_newuserdata(L, sizeof(EVP_CIPHER_CTX));
      |                                                                     ^~~~~~~~~~~~~~
src/lcrypto.c: In function ‘decrypt_fdecrypt’:
src/lcrypto.c:561:20: error: storage size of ‘c’ isn’t known
  561 |     EVP_CIPHER_CTX c;
      |                    ^
src/lcrypto.c: In function ‘hmac_pnew’:
src/lcrypto.c:604:57: error: invalid application of ‘sizeof’ to incomplete type ‘HMAC_CTX’ {aka ‘struct hmac_ctx_st’}
  604 |     HMAC_CTX *c = (HMAC_CTX *)lua_newuserdata(L, sizeof(HMAC_CTX));
      |                                                         ^~~~~~~~
src/lcrypto.c: In function ‘hmac_fnew’:
src/lcrypto.c:621:5: warning: implicit declaration of function ‘HMAC_CTX_init’; did you mean ‘HMAC_CTX_new’? [-Wimplicit-function-declaration]
  621 |     HMAC_CTX_init(c);
      |     ^~~~~~~~~~~~~
      |     HMAC_CTX_new
src/lcrypto.c: In function ‘hmac_clone’:
src/lcrypto.c:631:5: error: dereferencing pointer to incomplete type ‘HMAC_CTX’ {aka ‘struct hmac_ctx_st’}
  631 |     *d = *c;
      |     ^~
src/lcrypto.c: In function ‘hmac_gc’:
src/lcrypto.c:697:5: warning: implicit declaration of function ‘HMAC_CTX_cleanup’ [-Wimplicit-function-declaration]
  697 |     HMAC_CTX_cleanup(c);
      |     ^~~~~~~~~~~~~~~~
src/lcrypto.c: In function ‘hmac_fdigest’:
src/lcrypto.c:711:14: error: storage size of ‘c’ isn’t known
  711 |     HMAC_CTX c;
      |              ^
src/lcrypto.c: In function ‘sign_pnew’:
src/lcrypto.c:747:61: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_MD_CTX’ {aka ‘struct evp_md_ctx_st’}
  747 |     EVP_MD_CTX *c = (EVP_MD_CTX *)lua_newuserdata(L, sizeof(EVP_MD_CTX));
      |                                                             ^~~~~~~~~~
src/lcrypto.c: In function ‘sign_fsign’:
src/lcrypto.c:828:20: error: storage size of ‘c’ isn’t known
  828 |         EVP_MD_CTX c;
      |                    ^
src/lcrypto.c: In function ‘verify_pnew’:
src/lcrypto.c:857:61: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_MD_CTX’ {aka ‘struct evp_md_ctx_st’}
  857 |     EVP_MD_CTX *c = (EVP_MD_CTX *)lua_newuserdata(L, sizeof(EVP_MD_CTX));
      |                                                             ^~~~~~~~~~
src/lcrypto.c: In function ‘verify_fverify’:
src/lcrypto.c:940:20: error: storage size of ‘c’ isn’t known
  940 |         EVP_MD_CTX c;
      |                    ^
src/lcrypto.c: In function ‘rand_pseudo_bytes’:
src/lcrypto.c:992:5: warning: ‘RAND_pseudo_bytes’ is deprecated [-Wdeprecated-declarations]
  992 |     return rand_do_bytes(L, RAND_pseudo_bytes);
      |     ^~~~~~
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/err.h:13,
                 from src/lcrypto.c:7:
/usr/include/openssl/rand.h:44:1: note: declared here
   44 | DEPRECATEDIN_1_1_0(int RAND_pseudo_bytes(unsigned char *buf, int num))
      | ^~~~~~~~~~~~~~~~~~
src/lcrypto.c: In function ‘pkey_generate’:
src/lcrypto.c:1064:9: warning: ‘RSA_generate_key’ is deprecated [-Wdeprecated-declarations]
 1064 |         RSA *rsa = RSA_generate_key(key_len, RSA_F4, NULL, NULL);
      |         ^~~
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/err.h:13,
                 from src/lcrypto.c:7:
/usr/include/openssl/rsa.h:234:1: note: declared here
  234 | DEPRECATEDIN_0_9_8(RSA *RSA_generate_key(int bits, unsigned long e, void
      | ^~~~~~~~~~~~~~~~~~
src/lcrypto.c:1074:9: warning: ‘DSA_generate_parameters’ is deprecated [-Wdeprecated-declarations]
 1074 |         DSA *dsa = DSA_generate_parameters(key_len, NULL, 0, NULL, NULL, NULL, NULL);
      |         ^~~
In file included from /usr/include/openssl/e_os2.h:13,
                 from /usr/include/openssl/err.h:13,
                 from src/lcrypto.c:7:
/usr/include/openssl/dsa.h:117:1: note: declared here
  117 | DEPRECATEDIN_0_9_8(DSA *DSA_generate_parameters(int bits,
      | ^~~~~~~~~~~~~~~~~~
src/lcrypto.c: In function ‘pkey_to_pem’:
src/lcrypto.c:1164:27: error: dereferencing pointer to incomplete type ‘struct evp_pkey_st’
 1164 |     if (private && pkey_st->type == EVP_PKEY_DSA)
      |                           ^~
src/lcrypto.c: In function ‘pkey_tostring’:
src/lcrypto.c:1277:60: error: dereferencing pointer to incomplete type ‘EVP_PKEY’ {aka ‘struct evp_pkey_st’}
 1277 |     sprintf(buf, "%s %s %d %p", LUACRYPTO_PKEYNAME, (*pkey)->type == EVP_PKEY_DSA ? "DSA" : "RSA", EVP_PKEY_bits(*pkey), pkey);
      |                                                            ^~
src/lcrypto.c: In function ‘seal_pnew’:
src/lcrypto.c:1299:46: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_CIPHER_CTX’ {aka ‘struct evp_cipher_ctx_st’}
 1299 |     c->ctx = (EVP_CIPHER_CTX *)malloc(sizeof(EVP_CIPHER_CTX));
      |                                              ^~~~~~~~~~~~~~
In file included from src/lcrypto.c:8:
src/lcrypto.c: In function ‘seal_tostring’:
src/lcrypto.c:1320:81: error: dereferencing pointer to incomplete type ‘EVP_CIPHER_CTX’ {aka ‘struct evp_cipher_ctx_st’}
 1320 |     sprintf(s, "%s %p %s", LUACRYPTO_SEALNAME, (void *)c, EVP_CIPHER_name(c->ctx->cipher));
      |                                                                                 ^~
src/lcrypto.c: In function ‘seal_fseal’:
src/lcrypto.c:1400:20: error: storage size of ‘ctx’ isn’t known
 1400 |     EVP_CIPHER_CTX ctx;
      |                    ^~~
src/lcrypto.c: In function ‘open_pnew’:
src/lcrypto.c:1483:46: error: invalid application of ‘sizeof’ to incomplete type ‘EVP_CIPHER_CTX’ {aka ‘struct evp_cipher_ctx_st’}
 1483 |     c->ctx = (EVP_CIPHER_CTX *)malloc(sizeof(EVP_CIPHER_CTX));
      |                                              ^~~~~~~~~~~~~~
src/lcrypto.c: In function ‘open_fopen’:
src/lcrypto.c:1602:20: error: storage size of ‘ctx’ isn’t known
 1602 |     EVP_CIPHER_CTX ctx;
      |                    ^~~

Error: Failed installing dependency: https://raw.githubusercontent.com/rocks-moonscript-org/moonrocks-mirror/master/luacrypto-0.3.2-2.src.rock - Build error: Failed compiling object src/lcrypto.o
`
So how I solve this problem?

**Comment by MobtgZhang** [comment]
And while installing luaossl ,the compiler reports the error:
`Error: Could not find library file for CRYPTO
  No file libcrypto.a in /usr/lib
  No file libcrypto.so in /usr/lib
  No file matching libcrypto.so.* in /usr/lib
You may have to install CRYPTO in your system and/or pass CRYPTO_DIR or CRYPTO_LIBDIR to the luarocks command.
Example: luarocks install luaossl CRYPTO_DIR=/usr/local`

But I have installed the openssl ,the operating system is ubuntu19.10:
sudo apt install openssl libssl-dev

**Comment by GeoffRichards** [comment]
I'm on Ubuntu 16.04.6, and the library it uses comes from the package libssl1.0.0.

**Comment by coderforlife** [comment]
@MobTgZhang @GeoffRichards What does any of this have to do with this pull request?

**Review by damencho — APPROVED** [review]

**Comment by saghul** [comment]
Fixed in https://github.com/jitsi/luajwtjitsi
